src/ui/reducers/sources.ts

```diff
@@ -328,6 +328,27 @@ export function getHasSiblingOfSameName(state: UIState, source: MiniSource) {
   return state.experimentalSources.sourcesByUrl[source.url]?.length > 0;
 }
 
+export const getCanonicalSourceFromEntities = (
+  detailsEntities: Dictionary<SourceDetails>,
+  sd: SourceDetails
+) => {
+  const canonicalSource = detailsEntities[sd.canonicalId];
+  return canonicalSource ?? sd;
+};
+
+export const getCanonicalSource = (state: UIState, sd: SourceDetails) => {
+  return getCanonicalSourceFromEntities(state.experimentalSources.sourceDetails.entities, sd);
+};
+
+export const getCanonicalSourceForUrl = (state: UIState, url: string) => {
+  const sd = getSourceByUrl(state, url);
+  if (!sd) {
```

Comment:
I would also maybe add logging here. It's a bug if we're trying to open a source for an invalid URL.